### ISSUE TITLE
getOutputPaths() as well as getPages() Event Listeners do not deliver paths for generated html files.

### ISSUE THREAD

**Issue by nicolaskopp** [tighten/jigsaw#602]
As the title says.

if I start with an empty jigsaw project (no template), normally one index.html file is generated on build. Using said `getOutputPaths()` or `getPages()` Event Lsitener in boostrap.php does not deliver a collection entry for the generated index.html file, instead, the first array entry is empty, followed by the paths for the flatfiles (like png, css, js and so on).

I found this behaviour when I tried to use this function for implementing an HTML minifier for jigsaw: https://github.com/tighten/jigsaw/pull/116#issuecomment-993916984

If you need fruther info, please let me  know. Installed jigsaw yesterday, so I think it's a pretty recent version.

oh, and if you have built-in nice error logging function, pleeeeeeeease let me know, for now I am using error_log on build and it feels so so so wrong. Documentation about deugging and logging would be really helpful :)

**Comment by bakerkretzmar** [comment]
What exactly do they return for you? Are you on Windows?

**Comment by bakerkretzmar** [comment]
Gotcha, thanks. `/index.html` should be the only one that's ever missing, its relative 'pretty' path is `/` which gets trimmed to `''`.

For now you should be able to safely assume that there will always be exactly one item in the output paths array that is an empty string, and that path is actually `/index.html`, and likewise that there will be one item in the pages array with a key that's just an empty string, and it's actually the page at `/index.html`. Not sure if this is intentional, I'll find out and follow up.

To get a better look at what's in those collections, you can `dump()` them so they show up right in the build output and are highlighted:

```php
$events->afterBuild(fn ($jigsaw) => dump($jigsaw->getOutputPaths()));
$events->afterBuild(fn ($jigsaw) => dump($jigsaw->getPages()));
```

**Comment by nicolaskopp** [comment]
`dump` is a GameChanger!!! Makes life so much easier!

Okay, I will investigate and see if I can work around this bug and implement my Minifier in a more convenient way :) 

>Not sure if this is intentional, I'll find out and follow up.

doc states:

_"getPages() (**afterBuild** only)_

_Returns a collection of **all** output files that were generated. "_

So..I guess...Not intentional? :)

Thanks a lot!

**Comment by nicolaskopp** [comment]
one more thing, dump() does not output logs to the build outout for me, but maybe I am mistaken where to look - can you please point me into the right direction on where to find the results of dump(), please?

**Comment by nicolaskopp** [comment]
using something along the lines of this now and hope it does not break :)

```PHP
        $build_files = $jigsaw->getOutputPaths();
        $build_files->each(function ($item) use ($jigsaw) {

            if (gettype($item) == "string") {
                if (strlen($item) == 0) {
                    $item = "/index.html";
                }

                if (str_contains($item, ".html")) {
                    minify($item, $jigsaw);
                }
            }
        });
```

**Comment by bakerkretzmar** [comment]
> one more thing, dump() does not output logs to the build outout for me, but maybe I am mistaken where to look - can you please point me into the right direction on where to find the results of dump(), please?

It should just show up in your terminal, wherever you're running the build.

**Comment by nicolaskopp** [comment]
>It should just show up in your terminal, wherever you're running the build.

it does not :-/ again, Mac OS, also recently upgraded to PHP 7.4, using zsh terminal. 

I am getting this:

bootstrap.php:

```PHP
$events->afterBuild(fn ($jigsaw) => dump($jigsaw->getOutputPaths()));
```

Terminal output:

<img width="329" alt="Screen Shot 2021-12-16 at 11 18 57" src="https://user-images.githubusercontent.com/38662598/146353233-ea3a4c4d-2717-4192-b7f4-92f9b594fa47.png">

**Comment by bakerkretzmar** [comment]
It looks like there's other output missing there... does that build succeed? On a fresh install it should show the two lines you have there, then the dumps, then `Build time: ...` and `Site built successfully`.

**Comment by nicolaskopp** [comment]
Hope this helps:

### Scenario 1:

_bootstrap.php_

```PHP
<?php

use TightenCo\Jigsaw\Jigsaw;

/** @var $container \Illuminate\Container\Container */
/** @var $events \TightenCo\Jigsaw\Events\EventBus */

$events->afterBuild(fn ($jigsaw) => dump($jigsaw->getOutputPaths()));


```

_Results:_

**build**: builds normally, all files are built and copied as expected, with the following output:

<img width="329" alt="Screen Shot 2021-12-16 at 11 18 57" src="https://user-images.githubusercontent.com/38662598/146447950-4d6033bc-eeaf-4950-8f46-3862824fc667.png">

**serve**: does not start server. Hangs.

<img width="306" alt="Screen Shot 2021-12-16 at 21 58 14" src="https://user-images.githubusercontent.com/38662598/146448051-c5843562-b4b1-480e-b11d-0078d9aaa8d6.png">

### Scenario 2:

_bootstrap.php_

```PHP
<?php

use TightenCo\Jigsaw\Jigsaw;

/** @var $container \Illuminate\Container\Container */
/** @var $events \TightenCo\Jigsaw\Events\EventBus */


```

_Results:_

**build**: builds normally, all files are built and copied as expected, with the following output:

<img width="293" alt="Screen Shot 2021-12-16 at 21 59 16" src="https://user-images.githubusercontent.com/38662598/146448235-f4ca93ef-a6cd-4c18-85a6-8fcf1c99d396.png">

**serve**: Starts server, with the following output:
<img width="641" alt="Screen Shot 2021-12-16 at 22 00 31" src="https://user-images.githubusercontent.com/38662598/146448299-188aafca-1886-4556-aa62-6b669b7b8029.png">

**Comment by bakerkretzmar** [comment]
Weird... are you calling `dump()` anywhere else, or manually `var_dump`ing or `echo`ing other stuff?

**Comment by nicolaskopp** [comment]
nope, fresh install. If you want, we can set up a quick call next week and you can investigate my setup remotely :) 

If there is any other branch or thing I can test, please let me know.
